win,fs: handle AppExecLink reparse points in fstat

11soda11 opened the pull request:
Currently, libuv throws EACCES when trying to fstat AppExecLink reparse points, because Win32's CreateFileW can't resolve them on its own.

Use `fs__readlink_handle` as a fallback for reparse points that CreateFileW can't handle.

Fixes this nodejs issue: https://github.com/nodejs/node/issues/36790

This was already attempted (https://github.com/libuv/libuv/pull/4663) but the author of that pr stopped working on it.
I believe this is what was intended in https://github.com/libuv/libuv/pull/4663#pullrequestreview-2524405802 and https://github.com/libuv/libuv/pull/4663#pullrequestreview-2573143548 but correct me if im wrong.

Commit:
win,fs: handle AppExecLink reparse points in fstat

Currently, libuv throws EACCES when trying to fstat AppExecLink
reparse points, because Win32's CreateFileW can't resolve them on
its own.

Use `fs__readlink_handle` as a fallback for reparse points
that CreateFileW can't handle. Also change it to accept
a preallocated buffer.

vtjnash reviewed (approved):
I didn't thoroughly review, but everything looks correct at a high level

11soda11 commented on src/win/fs.c:
```diff
@@ -315,8 +316,82 @@ static int fs__readlink_handle(HANDLE handle,
+
+/* CreateFile wrapper. This uses our own readlink logic if windows can't do it. */
+static HANDLE fs__create_file(WCHAR* path, DWORD desired_access, int do_lstat) {
+  if (!do_lstat) {
```
I see. Should it then always be recursive? Because like this it would only ever be recursive up until the first symlink that CreateFile can handle by itsself.

vtjnash commented on src/win/fs.c:
```diff
@@ -315,8 +328,137 @@ static int fs__readlink_handle(HANDLE handle,
+                         FILE_FLAG_BACKUP_SEMANTICS,
+                         NULL);
+
+    if (handle != INVALID_HANDLE_VALUE)
```
Is there any useful error to check here to know if it was a failed reparse point operation? Seems like a python developer said it is ERROR_CANT_ACCESS_FILE, and should not be handled https://github.com/python/cpython/issues/104336#issuecomment-1540878585
```suggestion
    if (handle != INVALID_HANDLE_VALUE || GetLastError() != ERROR_CANT_ACCESS_FILE)
```

Although that also implies our mapping of it to EACCES was wrong, and it should instead be mapped to
>       ENOENT A directory component in pathname does not exist or is a dangling symbolic link.

vtjnash replied:
Yeah, even if it is inaccurate, we shouldn't change that errno mapping now

11soda11 commented on src/win/fs.c:
```diff
@@ -315,8 +328,137 @@ static int fs__readlink_handle(HANDLE handle,
+        if (need_free)
+          uv__free(target_path);
+
+        need_free = 0;
```
target_path is equal to path here, unless were in the first iteration of the loop, in which case need_free is always zero. For the same reason the code you provided produces a UAF. It only makes sense to free on UV_ENOBUFS here because that is the only time we will allocate a new buffer.

11soda11 commented:
I added the suggested changes + made a small change in `fs__create_file` (line 439 + 460) to skip an unreachable if statement.
Also added the test, which turned out to be a oneliner.

vtjnash reviewed (approved):
Looks good to merge, unless anyone else wants to review

bnoordhuis commented on src/win/fs.c:
```diff
@@ -315,8 +328,137 @@ static int fs__readlink_handle(HANDLE handle,
+    *w_target_wtf8_len_ptr = uv_utf16_length_as_wtf8(w_target, w_target_len);
+
+  w_target_size = w_target_len + 1;
+  if (w_target_ptr) {
```
Minor stylistic issue but the libuv house style is to not let pointers auto-coerce to booleans. That is...
```suggestion
  if (w_target_ptr != NULL) {
```
...here and elsewhere.

bnoordhuis commented on src/win/fs.c:
```diff
@@ -315,8 +328,137 @@ static int fs__readlink_handle(HANDLE handle,
+         *w_target_size_ptr = w_target_size;
+
+       SetLastError(ERROR_OUTOFMEMORY);
+       return -1;
```
This is kind of nasty. Do `return UV_ENOMEM;` instead.

bnoordhuis commented on src/win/fs.c:
```diff
@@ -315,8 +328,137 @@ static int fs__readlink_handle(HANDLE handle,
+  if (w_target_size_ptr)
+    *w_target_size_ptr = w_target_len;
+
+
```
```suggestion
```

bnoordhuis commented on src/win/fs.c:
```diff
@@ -315,8 +328,137 @@ static int fs__readlink_handle(HANDLE handle,
+      }
+    }
+
+    CloseHandle(handle);
```
Also set `handle = INVALID_HANDLE_VALUE`, otherwise this function returns a closed handle when it drops out of the loop.

11soda11 replied:
very good catch, thx

bnoordhuis commented on src/win/fs.c:
```diff
@@ -315,8 +328,137 @@ static int fs__readlink_handle(HANDLE handle,
+    path = target_path;
+  }
+
+
```
```suggestion
```

bnoordhuis reviewed (commented):
At least one serious bug in `fs__create_file()` (bad handle return) and the loop there looks really dubious. It's trying to walk symbolic links manually, right? Why?

It's not obvious from just looking at the code, and it screams HERE BE TOCTOU BUGS to the casual reader.

11soda11 commented:
> At least one serious bug in `fs__create_file()` (bad handle return) and the loop there looks really dubious. It's trying to walk symbolic links manually, right? Why?

Fixed the bug and implemented all the other suggestions. The reason its walking through symbolic links manually is because windows will not do that by itself for AppExecLinks.

vtjnash commented on src/win/fs.c:
```diff
@@ -315,8 +328,132 @@ static int fs__readlink_handle(HANDLE handle,
+  size_t target_path_size;
+  WCHAR target_path_scratch[512];
+  WCHAR* target_path = target_path_scratch;
+  size_t target_buf_size = sizeof(target_path_scratch);
```
Technically syntax (the `()` are optional!)

11soda11 replied:
But would it really violate the style guide if its compile time known? I know it needs to be divided by the size of WCHAR nonetheless, but that still seems a lot better than hardcoding the number twice

11soda11 replied:
Ah i see, ill use ARRAY_SIZE then. updated it again
